# Exercícios

## 1. Execute as células abaixo e explique os resultados

In [7]:
class A:
    def f(self):
        print('f of A')
        
    def g(self):
        print('g of A')
        self.f()

a = A()
a.f()
a.g()

f of A
g of A
f of A


a = A() 
criamos um objeto a.

a.f()
acessamos o método definido na classe com a notação de ponto. Assim temos o resultado 'f of A'

a.g()
acessamos o método definido na classe com a notação de ponto. Assim temos o resultado 'g of A' e após isso chamamos a função f() com self.f(). 

In [8]:
class B:
    def __init__(self, x):
        print(f'New B with {x}')
        self.x = x
        
    def f(self, y):
        print(f'f for a B with {self.x}')
        return self.x + y
    
b1 = B(4)
b2 = B(7)
print(f'{b1.f(10)}')
print(f'{b2.f(20)}')

New B with 4
New B with 7
f for a B with 4
14
f for a B with 7
27


b1 = B(4)
criamos objeto b1 em que passamos 4 como parâmetro. Ao criar um objeto o método __init__ é chamado e o parâmetro passado durante a criação é passado para esse método assim temos o resultado 'f New B with 4'. 

b2 = B(7)
criamos objeto b1 em que passamos 4 como parâmetro. Ao criar um objeto o método __init__ é chamado e o parâmetro passado durante a criação é passado para esse método assim temos o resultado 'f New B with 7'.

print(f'{b1.f(10)}')
acessamos o método definido na classe com a notação de ponto. Assim por meio do objeto b1 acessamos o método f com ponto passando o parâmetro 10. Então será printado 'f for a B with 4' que é o valor passado na criação do objeto b1 como também a soma self.x + y (10 + 4).

print(f'{b2.f(20)}')
acessamos o método definido na classe com a notação de ponto. Assim por meio do objeto b2 acessamos o método f com ponto passando o parâmetro 20. Então será printado 'f for a B with 7' que é o valor passado na criação do objeto b2 como também a soma self.x + y (20 + 7).

In [9]:
class C:
    def f(self):
        print('f of C')
    def g(self):
        print('g of C')
        self.f()

class D(C):
    def f(self):
        print('f of D')
              
c = C()
d = D()
c.f()
c.g()
d.f()
d.g()

f of C
g of C
f of C
f of D
g of C
f of D


c = C()
criamos o objeto c sem passar nenhum parâmetro.

d = D()
criamos o objeto d sem passar nenhum parâmetro. Vale destacar que a criação da classe D usa herança para herdar o código existente da classe C.

c.f()
acessamos o método definido na classe C com a notação do ponto. Assim por meio do objeto c acessamos o método f com o ponto sem passar nenhum parâmetro. Então temos o resultado 'f of C'.

c.g()
acessamos o método definido na classe C com a notação do ponto. Assim por meio do objeto c acessamos o método g com o ponto sem passar nenhum parâmetro. Então temos o resultado 'g of C'. Mas também temos a chamada do método f logo teremos mais um resultado 'f of C'. 

d.f()
acessamos o método definido na classe D com a notação do ponto. Assim por meio do objeto d acessamos o método f com o ponto sem passar nenhum parâmetro. Então temos o resultado 'f of D'.

d.g()
acessamos o método definido na classe D com a notação do ponto. Assim por meio do objeto d acessamos o método g herdado de C com o ponto sem passar nenhum parâmetro. Assim é solicitado a impressão de 'g of C' e logo abaixo é chamado o método f e assim temos o resultado 'f of D'

In [10]:
class E:
    def __init__(self, x):
        self.x = x
    def f(self):
        return self.x

class F(E):
    def f(self):
        return 2 * self.x
    
e = E(12)
f = F(22)
print(e.f())
print(f.f())

12
44


e = E(12)
criamos o objeto e passando o parâmetro 12.

f = F(22)
criamos o objeto f passando o parâmetro 22.

print(e.f())
acessamos o método f definido na classe E com a notação de ponto. Assim temos criado o valor da variável x guardado em self.x no __init__ então nosso resultado será o valor passado como parâmetro na criação do objeto. Logo nosso resultado será 12.

print(f.f())
acessamos o método f definido na classa F em que vamos multiplicar por dois o valor herdado na criação do objeto da classe E (22). Logo nosso resultao será 44. 


In [11]:
class G:
    def __init__(self, x):
        self.x = x
    def f(self):
        return self.x
    
class H(G):
    def __init__(self, x):
        self.x = 3 * x
        
g = G(15)
h = H(35)
print(g.f())
print(h.f())

15
105


g = G(15)
criamos um objeto g passando 15 como parâmetro.

h = H(35)
criamos um objeto h passando 35 como parâmetro. Vale destacar que a classe H herda os todos métodos de G.

print(g.f())
acessamos o método f definido na classe G com a notação de ponto. Assim na criação da classe G é passado para o método __init__ o valor de 15 guardado em self.x e o método f quando chamado apenas retorna esse valor. Assim temos o resultado 15.

print(h.f())
na criação da classe H temos a reedefinição do método __init__ em que o valor passado como parâmetro é multiplicado por 3. Como a classe H não tem o método f, mas herda a classe G que tem o método f então esse método é chamado para retornar o valor guardado no __init__ que foi redefinido. Então temos o resultado 3 * 35 == 105.

In [16]:
class I:
    def __init__(self, x):
        self.x = x
    def f(self):
        return self.x
    def g(self):
        return 10 * self.f()
    
class J(I):
    def __init__(self, y):
        self.y = y
    def g(self):
        return 30 * self.f()
    
i = I(3)
j = J(11)
print(i.f())
print(i.g())
print(j.g())

3
30


AttributeError: 'J' object has no attribute 'x'

i = I(3)
criamos o objeto i passando como parãmetro 3. A função __init__ é iniciada e guarda o valor do parâmetro na variável self.x. 

j = J(11)
criamos o objeto j passando como parãmetro 11.

print(i.f())
acessamos o método f() definido na classe I com a notação de ponto. Assim, é retornado o valor de x que é o parâmetro passado na criação do objeto i já que o __init__ (que é a função iniciada na criação do objeto) de I guarda esse valor na varável self.x 

print(i.g())
acessamos o médoto g() definido na classe I com a notação de ponto. Este retorna a multiplicação por dez do método f() e do item anterior temos que f() retorna 3. Assim o resultado será 30.

In [1]:
class AA:
    def __init__(self, x):
        self.x = x
    def f(self, y):
        return self.x + y
    def g(self, y):
        return 2 * self.f(y)
    
class BB(AA):
    def f(self, y):
        return self.x - y

class CC(AA):
    def g(self, y):
        return 3 * self.f(y)
    
class DD(BB):
    def g(self, y):
        return 5 * self.f(y)
    
class EE(CC):
    def f(self, y):
        return self.x + 2 * y
    
class FF(CC):
    def __init__(self, x):
        CC.__init__(self, 5 * x)
        
aa = AA(7)
bb = BB(11)
cc = CC(13)
dd = DD(17)
ee = EE(19)
ff = FF(23)
print(aa.g(2))
print(bb.g(2))
print(cc.g(2))
print(dd.g(2))
print(ee.g(2))
print(ff.g(2))

18
18
45
75
69
351


criamos os seguintes objetos passando os parâmetros em parenteses: aa = AA(7), bb = BB(11), cc = CC(13), dd = DD(17), ee = EE(19) e ff = FF(23).

print(aa.g(2)) acessamos o método g() definido na classe AA com a notação de ponto. O __init__ da classe guarda o valor passado como parâmetro na criação do objeto na variável self.x = 7. O método g() multiplica por dois o valor da chamada do método f(y). O método f() soma a essa variável um y passado ao método f(y): self.x + y. Assim o resultado será 18.

print(bb.g(2)) antes vale destacar que a classe BB herda a classe AA como também redefine o método f(). Acessamos o método g() com a notação de ponto. O método g() multiplica por dois o valor da chamada do método f(y). O método f() subtrai o valor passado como parâmetro na criação do objeto pelo valor passado para o método: 11 - 2. Assim o resultado final será 18.

print(cc.g(2)) a classe CC também herda AA e redefine o método g(). Acessamos o método g() com a notação de ponto. Este multiplica por 3 a chamada de f(y). O método f(y) soma o valor passado como parâmetro na criação do objeto cc com o valor passado a este método: 13 + 2. Assim o resultado final será 45.

print(dd.g(2)) a classe DD herda BB e AA, pois BB herda AA como também redefine o método g() de AA. Acessamos o método g(2) passando um parâmetro com a notação de ponto. Este método multiplica por 5 o retorno da chamada do método f(2). O método f() definido na classe BB pega o valor passado como parâmetro na criação do objeto dd(17) e subtrai pelo passado pelo método f(2) então 17 - 2 = 15. Assim o resultado final é 75.

print(ee.g(2)) a classe EE herda CC e AA, pois CC herda AA como também redefine o método f() de AA. Acessamos o método g() passando o parâmetro 2 com a notação de ponto. Este método multiplica por 3 o retorno da chamada de f(2). O método agora redefinido soma o valor passado como parâmetro na criação do objeto EE(19) com a multiplicação de 2 pelo valor passado ao método g(2) então 19 + 2X2 = 23. assim o resultado final fica 69.

print(ff.g(2)) a classe FF herda CC e AA, pois CC herda AA como também define um __init__ em que chama o __init__ da classe CC passando o parâmetro que foi usado na criação do objeto ff = FF(23) multiplicado por 5 então 115. Já o __init__ da classe CC é o da classe AA por herença. Este guarga na variável self.x o o valor self.x = 115. Acessamos o método g com a notação de ponto passando o parãmetro 2. O método g(2) retorna a multiplicação 3 pela chamada de f(2). Este método soma self.x + y então 115 + 2 = 117. Então o resultado final é 351.  

## 2. Implemente a classe pedida

Temos a seguinte classe disponível que implementa um ponto num espaço tridimensional:

In [10]:
class Point:
    def __init__(self, x, y, z):
        self._coords = x, y, z
    def x(self):
        return self._coords[0]
    def y(self):
        return self._coords[1]
    def z(self):
        return self._coords[2]
    
class Particle(Point):
    def __init__(self, x, y, z, m):
        self.m = m
        Point.__init__(self, x, y, z)
    def mass(self):
        return self.m

Usando herança, crie uma classe denominada `Particle`, que tem além da posição da partícula a sua massa; a massa é fornecida junto com as coordenadas na criação do objeto. Essa classe deve ter um método denominado `mass` que retorna o valor da massa da partícula.

Veja o exemplo de uso abaixo:

In [11]:
p1 = Particle(1.0, 2.0, -1.0, 100.) # partícula em (1, 2, -1) com massa 100
print(f'A particle with mass {p1.mass()} was detected at ({p1.x()}, {p1.y()}, {p1.z()})')

A particle with mass 100.0 was detected at (1.0, 2.0, -1.0)
